### ISSUE TITLE
Speech2Text 모델 ONNX 변환 관련 문의

### ISSUE THREAD

**Issue by Barleysack** [hchung12/espnet-asr#4]
안녕하세요.

좋은 사전학습 모델 및 코드 감사드립니다.
이 모델의 Inference 속도 향상을 위해 onnx로 바꿔 tensorRT로 변환해 구동해보려고 하는데, 
Speech2Text 모델을 변환하려니 nn.Module의 객체 여야 하기에 토치로 간단히 래핑하여 변환을 시도했습니다.

그랬더니 Cannot insert a Tensor that requires grad as a constant. Consider making it a parameter or input, or detaching the gradient.
라는 오류 메시지와 함께 변환이 안되더라구요.

찾아보니 모델 내부에 커스텀 연산이 들어있으면 그렇게 될 가능성이 있다고 하기에, 내부의 빔서치 알고리즘이 문제인가 싶어
지금의 Speech2Text 모델을 여러 조각으로 분리하여 이어 붙이려고 합니다. 

혹시 개발자님께서는 onnx 변환을 시도해보셨나요? 
그렇다면 어떤 방식으로 진행하셨을지 알 수 있을까요?

감사합니다.

현재의 변환 코드는 다음과 같습니다. 

``` 
    test_model.eval()
    SOUND_PATH = PATH
    preprocess_fn=ASRTask.build_preprocess_fn(model.asr_train_args, False)
    sound= AudioSegment.from_wav(SOUND_PATH)

    sound = pcm2float(sound.get_array_of_samples())
    sound= preprocess_fn('1',{'speech':sound})
    sound =torch.from_numpy(sound['speech'])
    onnx_model = torch.onnx.export(test_model,sound,"letussleep.onnx",input_names=['input'],
                    output_names=['output'],export_params=True) 
```

이때의 test_model은 다음과 같습니다.

``` 
class TestWrapperModel(nn.Module):
    def __init__(self, model):
        super(TestWrapperModel,self).__init__()
        self.s2t = model
        
    def forward(self,x):
        x= self.s2t(x)
        
        return 
```
이때의 model 은 작성해주신 코드의 Speech2Text 객체입니다.

**Comment by Barleysack** [comment]
이후 인코더와 빔서치로 나누어서 저장해보려고 했지만....인코더는 저장이 되지만 빔서치는 내부의 ctc와 디코더, 그리고 빔 서치 자체 내부 연산등으로 인해 변환이 안되네요... 혹시 시도해 보셨나요 ㅠㅠ

**Comment by hchung12** [comment]
안녕하세요, ONNX로 변환을 시도해보지 않았서 도움을 드리기는 어려울 것 같습니다. espnet  공식 홈페이지에서 ONNX관련 issue  (https://github.com/espnet/espnet/issues?q=onnx) 를 참고해 보시면 도움이 될지도 모르겠습니다. 감사합니다.

**Comment by Barleysack** [comment]
감사합니다. 공식 홈페이지의 컨트리뷰터들은 내부의 동적 연산들로 인해 ONNX로 바꿀 수 없다고, 아마 그걸 바꾸는 것과 코드를 새로 쓰는게 거의 비슷할 것이라는 입장이네요 :) 제법 known issue인듯 합니다. 혹시 그렇다면 개발자님께서는 최적화를 위해 시도해보신 점이 있나요? 
빠른 답변 감사드립니다.

**Comment by hchung12** [comment]
안녕하세요, 코드 엔지니어링을 통한 최적화 작업을 진행한 경험은 없습니다.
다만 신경망 모델을 low rank factorization을 통해 파라메타 수와 계산량을 줄이는 연구를 진행한 적은 있는데,
이는 ALBERT(https://arxiv.org/pdf/1909.11942.pdf)의 Factorized embedding parameterization 방식과 유사하다고 할 수 있습니다.
감사합니다.

**Comment by Barleysack** [comment]
답변 감사드립니다. 
ONNX 가능한 모듈로 새로 짜올려서 4개 데이터셋에 대해 학습을 시키던지,
지금의 속도에 만족하던지 해야겠네요 :)
감사합니다!
